dali/operators/audio/preemphasis_filter_op.cu

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/preemphasis_filter_op.h"
+#include <vector>
+
+namespace dali {
+
+namespace detail {
+
+template <typename OutputType, typename InputType>
+struct SampleDescriptor {
+  const InputType *in;
+  OutputType *out;
+  float coeff;
+  int64_t size;
+};
+
+template <typename OutputType, typename InputType>
+void __global__ PreemphasisFilterKernel(const SampleDescriptor<OutputType, InputType> *samples) {
+  const auto &sample = samples[blockIdx.y];
+  int64_t block_size = blockDim.x;
+  int64_t block_start = block_size * blockIdx.x;
+  int64_t grid_stride = block_size * gridDim.x;
+
+  int64_t k = block_start + threadIdx.x;
+  if (k >= sample.size)
+    return;
+
+  sample.out[k] = sample.in[k] - sample.coeff * sample.in[cuda_max(0l, k-1)];
+  k += grid_stride;
+  for (; k < sample.size; k += grid_stride)
+    sample.out[k] = sample.in[k] - sample.coeff * sample.in[k-1];
+}
+
+}  // namespace detail
+
+class PreemphasisFilterGPU : public PreemphasisFilter<GPUBackend> {
+ public:
+  explicit PreemphasisFilterGPU(const OpSpec &spec) : PreemphasisFilter<GPUBackend>(spec) {}
+  void RunImpl(workspace_t<GPUBackend> &ws) override;
+
+ private:
+  template <typename OutputType, typename InputType>
+  void RunImplTyped(workspace_t<GPUBackend> &ws);
+
+  Tensor<GPUBackend> scratchpad_;
```

Comment:
Nitpick - I don't know if we it is the best name as it has nothing to do with the `scratchpad` entity.

Reply:
done